### DISCUSSION TITLE
[V6] Singleton per-request?

### DISCUSSION THREAD

**Discussion by holoyan** [adonisjs/core#4498] [Help] [answered]
Is there a way to bind singleton per-request?

```typescript

// scope.ts


export default class Scope {
 value = 0
}


// provider.ts

this.app.container.singleton('scope', async () => {
  return new Scope()
})


// routes.ts

// first call this
router.get('/', async () => {

  const scope = await app.container.make('scope')
  scope.value = 5

})

// then call localhost/test2
router.get('/test2', async () => {

  const scope = await app.container.make('scope')
  console.log(scope.value) // curentlly it will print 5 but I want to have 0
})



```

Is there a way we can make singleton per-request through IoC container?

Update:

I know how to do that using middleware, I'm just wondering if there is a way to do like in the [nest](https://docs.nestjs.com/fundamentals/injection-scopes)

**Comment by thetutlage** [comment]
Middleware is the way to go. As per me, its a bit more explicit and easy to reason about, since you can clearly understand the flow how and when a binding is registered in the container.

Lemme know if you have any specific struggles with the middleware approach?

**Reply by holoyan** [reply]
```typescript

// scope.ts

export default class Scope {
 value = 0
}


```


So the goal is to have  `new` singleton object for ONLY current request lifesycle, which developers can invoke anywhere inside the project (controller, service class and so on...) and change `value` property.


And the reason why I don't like middleware approach is because I've to strictly bind my `Scope` object to the [request class](https://docs.adonisjs.com/guides/request#extending-request-class) and pass that request everywhere in the project

(Hope I didn't miss anything from the docs)


```typescript

Request.getter('scope', function (this: Request) {
  return new Scope()
})


```

```typescript
// MyService.ts


foo(ctx: HttpContext){

// Scope is kind of internal 'thing' and I don't thing it's good aproach to get it from the input
  ctx.request.input('scope').value = 5

}

```

Instead I wanted to have something like this

```typescript

// bind Scope

this.app.container.singletonPerHttpRequest('scope', async () => {
  return new Scope()
})

}

```

and then everywhere I can just get it from the container 

```typescript

// SomeService.ts

  const scope = await app.container.make('scope')
  scope.value = 5

```

**Reply by RomainLanz** [reply, marked as answer]
You can bind an injectable value in your middleware.

```ts
handle(ctx: HttpContext, next: NextFn) {
  const scope = new Scope()
  ctx.containerResolver.bindValue(Scope, scope)
}
```

Then:

```ts
@inject()
class MyService {
  constructor(private scope: Scope) {}
}
```

**Reply by holoyan** [reply]
Thanks @RomainLanz , it does the trick.

I'm just curious why multiple concurrent requests not overriding each other, would you mind to give a short explaination
> ctx.containerResolver.bindValue(Scope, scope)

I tought binding is 'just' a key-value pairs and becouse we have single instance of IoC container it will override to each other on every request.

**Reply by thetutlage** [reply]
The `ctx.containerResolver` is an isolated bucket of bindings. Think of it as follows.

```ts
createServer(() => {
  const ctx = createContext()
  ctx.containerResolver = {}

  await executeRequestHandlers(ctx)
})
```

So the values to put into `containerResolver` are isolated.

**Reply by holoyan** [reply]
Now that makes sense,

Thank you @thetutlage
